File: src/app/user/user.service.js
/*
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
class UserService {

  constructor($http, $rootScope, Constants) {
    'ngInject';
    this.$http = $http;
    this.$rootScope = $rootScope;
    this.usersURL = Constants.baseURL + 'users/';
    this.userURL = Constants.baseURL + 'user/';
  }

  list() {
    return this.$http.get(this.usersURL);
  }

  get(code) {
    return this.$http.get(this.usersURL + code);
  }

  listTeams(code) {
    return this.$http.get(this.usersURL + code + '/teams');
  }

  create(user) {
    return this.$http.post(this.usersURL, user);
  }

	search(query) {
		return this.$http.get(this.usersURL + "?query=" + query);
	}

	isUserInRoles(roles) {
    roles = (Array.isArray(roles)) ? roles : [roles];
    
	  if (!this.$rootScope.graviteeUser) {
	    return false;
	  }

	  if (this.$rootScope.graviteeUser && (!roles || roles.length == 0)) {
	    return false;
	  }

	  var rolesAllowed = false, that = this;
	  _.forEach(roles, function(role) {
	    _.forEach(that.$rootScope.graviteeUser.authorities, function(authority) {
	      if (authority.authority === role) {
	        rolesAllowed = true;
	        return;
	      }
	    });
	  });

	  return rolesAllowed;
	}

  current() {
    return this.$http.get(this.userURL);
  }

  login(user) {
    var req = {
      method: 'POST',
      url: this.userURL + 'login',
      headers: {
        'Authorization': "Basic " + btoa(user.username + ":" + user.password)
      }
    };
    return this.$http(req);
  }

  logout() {
    return this.$http.post(this.userURL + 'logout');
  }

  currentUserPicture() {
    return this.$http.get(this.userURL + this.$rootScope.graviteeUser.username + '/picture');
  }

  save(user) {
    return this.$http.put(this.userURL + this.$rootScope.graviteeUser.username + '/', {username: user.username, picture: user.picture});
  }
}

export default UserService;

File: gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/rest/api/service/UserService.java
/**
 * Copyright (C) 2015 The Gravitee team (http://gravitee.io)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.gravitee.rest.api.service;

import io.gravitee.common.data.domain.Page;
import io.gravitee.repository.management.api.search.UserCriteria;
import io.gravitee.rest.api.model.*;
import io.gravitee.rest.api.model.common.Pageable;
import io.gravitee.rest.api.model.configuration.identity.RoleMappingEntity;
import io.gravitee.rest.api.model.configuration.identity.SocialIdentityProviderEntity;
import io.gravitee.rest.api.service.common.ExecutionContext;
import io.gravitee.rest.api.service.common.JWTHelper.ACTION;
import java.util.*;

/**
 * @author David BRASSELY (david.brassely at graviteesource.com)
 * @author Nicolas GERAUD (nicolas.geraud at graviteesource.com)
 * @author Azize Elamrani (azize.elamrani at graviteesource.com)
 * @author GraviteeSource Team
 */
public interface UserService {
    UserEntity connect(ExecutionContext executionContext, String userId);

    UserEntity findById(ExecutionContext executionContext, String id, boolean defaultValue);

    default UserEntity findById(ExecutionContext executionContext, String id) {
        return findById(executionContext, id, false);
    }

    Optional<UserEntity> findByEmail(ExecutionContext executionContext, String email);

    UserEntity findByIdWithRoles(ExecutionContext executionContext, String id);

    UserEntity findBySource(ExecutionContext executionContext, String source, String sourceId, boolean loadRoles);

    Set<UserEntity> findByIds(ExecutionContext executionContext, Collection<String> ids);

    Set<UserEntity> findByIds(ExecutionContext executionContext, Collection<String> ids, boolean withUserMetadata);

    UserEntity create(ExecutionContext executionContext, NewExternalUserEntity newExternalUserEntity, boolean addDefaultRole);

    UserEntity update(ExecutionContext executionContext, String userId, UpdateUserEntity updateUserEntity);

    UserEntity update(ExecutionContext executionContext, String userId, UpdateUserEntity updateUserEntity, String newsletterEmail);

    Page<UserEntity> search(ExecutionContext executionContext, String query, Pageable pageable);

    Page<UserEntity> search(ExecutionContext executionContext, UserCriteria criteria, Pageable pageable);

    UserEntity register(ExecutionContext executionContext, NewExternalUserEntity newExternalUserEntity);

    UserEntity register(ExecutionContext executionContext, NewExternalUserEntity newExternalUserEntity, String confirmationPageUrl);

    UserEntity finalizeRegistration(ExecutionContext executionContext, RegisterUserEntity registerUserEntity);

    UserEntity finalizeResetPassword(ExecutionContext executionContext, ResetPasswordUserEntity registerUserEntity);

    UserEntity processRegistration(ExecutionContext executionContext, String userId, boolean accepted);

    PictureEntity getPicture(ExecutionContext executionContext, String id);

    void delete(ExecutionContext executionContext, String id);

    void resetPassword(ExecutionContext executionContext, String id);

    UserEntity resetPasswordFromSourceId(ExecutionContext executionContext, String sourceId, String resetPageUrl);

    Map<String, Object> getTokenRegistrationParams(
        ExecutionContext executionContext,
        UserEntity userEntity,
        String portalUri,
        ACTION action
    );

    Map<String, Object> getTokenRegistrationParams(
        ExecutionContext executionContext,
        UserEntity userEntity,
        String portalUri,
        ACTION action,
        String confirmationPageUrl
    );

    UserEntity create(ExecutionContext executionContext, NewPreRegisterUserEntity newPreRegisterUserEntity);

    UserEntity createOrUpdateUserFromSocialIdentityProvider(
        ExecutionContext executionContext,
        SocialIdentityProviderEntity socialProvider,
        String userInfo
    );

    void updateUserRoles(
        ExecutionContext executionContext,
        String userId,
        MembershipReferenceType referenceType,
        String referenceId,
        List<String> roleIds
    );

    void computeRolesToAddUser(
        ExecutionContext executionContext,
        String username,
        List<RoleMappingEntity> mappings,
        String userInfo,
        Set<RoleEntity> rolesToAddToOrganization,
        Map<String, Set<RoleEntity>> rolesToAddToEnvironments
    );
}
